### ISSUE TITLE
Use test set as additional training data

### ISSUE THREAD

**Issue by KDercksen** [KDercksen/hunter2_dbi#14]
I recently had this idea... It may seem a bit hacky but hear me out :D 

After training our model on the training set, we use it to make predictions about the test set. We can use these predictions as labels for the test set (although not 100% accurate), and use the labeled test set combined with the training data to further train the network. Assuming the classification error on the test set is already pretty good, this might provide the model with further clues about the nature of the data, and a second prediction might be even better. We can then use these improved labels to repeat the procedure, etc.

I don't know if this is worth looking into, I haven't done much research on it but intuitively it sounds like it may be promising (although a little cheaty, but not against the rules). Training the network in this way may take a bit longer than it already does, though... Anyone have any thoughts?

**Comment by stephandooper** [comment]
I don't think this would be a good idea.

One the one hand, you will be generating some incorrect labels, thus tainting the training procedure.
Especially if there are certain classes that the model has trouble with, then I expect this problem not to be solved using predictive labels (as the labels on the hard class(es)) will be the largest portion of incorrect classified examples.

The incorrect labels can then further harm training in quite a straightforward way: the incorrect labels can have impact for other classes as well. For example, dog breed with labels A,B, and C are the ones we are interested in. We find correct predictions for A,B but not for C, which ends up being labelled A. In the training, we then wrongly associate dog breed C to A, which can lead to be worse results for dog breed A as well. 

Finally, any analysis and exploration done on the test set from then on would be pointless as the correctly classified labels can just be dubbed training data, and will thus results in a "training error"  instead of a "test error".

**Comment by KDercksen** [comment]
I agree on your point that maybe the hard classes are gonna end up even worse using this method. That is an issue we'll first have to solve (#3); maybe with solid predictions across the board this will have a better use. However, I do think that the introduction of additional data (the test set) could help the model in the long run; the fact that we might be 'overfitting' to the test data is not necessarily a bad thing (with the goal of a high leaderboard rank in mind, anyway).

**Comment by SvenDH** [comment]
I think it is worth looking into this since it does not require that much work. Aside from that we could also try to find the hard to predict classes and oversample those in a second training run.

**Comment by KDercksen** [comment]
Oversampling is veeeery easy to implement (see #11), we could also use this for our initial training run even. I suspect it will help with accuracy immediately.

**Comment by KDercksen** [comment]
Did not increase accuracy. :(
